### ISSUE TITLE
Support Shared Access Signatures with Azure Blob Storage

### ISSUE THREAD

**Issue by memotype** [ncw/rclone#2118]
It would be nice if rclone supported SAS authorization instead of requiring the storage account's shared key. That way I can distribute a read-only SAS URL and use it with rclone.

**Comment by ncw** [comment]
Sounds like a reasonable request and it is probably [supported by the SDK](https://github.com/Azure/azure-sdk-for-go)

Fancy doing a bit of research, and linking to some docs?  Or even better sending a pull request?

**Comment by maetthu** [comment]
I'm not particularly versed enough in Go (yet) to contribute some code, but I'll pitch in with some info. For Blob Storage, there are two different kinds of SAS, one is an account level SAS which gives you access to all containers (or other storage products in this account if it's allowed, like tables or queues), the other one is a service level SAS which will give you access to a specific container (or file). 

For an account level SAS, you'll only need a different constructor - instead of https://godoc.org/github.com/Azure/azure-sdk-for-go/storage#NewClient you'll either need https://godoc.org/github.com/Azure/azure-sdk-for-go/storage#NewAccountSASClient, https://godoc.org/github.com/Azure/azure-sdk-for-go/storage#NewAccountSASClientFromEndpointToken or https://godoc.org/github.com/Azure/azure-sdk-for-go/storage#NewClientFromConnectionString, depending on which info you have. _NewClientFromConnectionString_ is probably the best one, since a connection string covers every connection method for account-level access:

* Access with an account key: 
`DefaultEndpointsProtocol=https;AccountName=<ACCOUNTNAME>;AccountKey=<ACCOUNTKEY>;BlobEndpoint=https://<ACCOUNTNAME>.blob.core.windows.net/;`
* With an SAS:
`SharedAccessSignature=sv=2017-04-17&ss=bt&srt=sco&sp=rl&st=2018-03-20T20%3A13%3A00Z&se=2018-03-21T20%3A13%3A00Z&sig=<SIGNATURE>;BlobEndpoint=https://<ACCOUNTNAME>.blob.core.windows.net/;`

IMHO it would be great to be able to either configure an Azure Blob Storage remote storage using the traditional way (account name, key, endpoint) for backwards compatibility or with a connection string, which covers everything supported by the SDK, including the former.

A service level SAS is a bit different. An SAS for a single file doesn't make much sense for rclone since it's just used to give access to a single file, but a container SAS might be useful. Using an SAS URL you can get a container instance using https://godoc.org/github.com/Azure/azure-sdk-for-go/storage#GetContainerReferenceFromSASURI and from there a Client instance again https://godoc.org/github.com/Azure/azure-sdk-for-go/storage#Container.Client

**Comment by ncw** [comment]
@maetthu thank you for the writeup - very helpful :-)

So it looks like adding access via `NewClientFromConnectionString` would be useful.  That looks pretty easy too.

To cover the use case of @memotype (the original poster) I'd need to add config for a SAS URL too if I'm understanding you correctly.

So the 3 config options would be
  * account/key
  * connection string
  * SAS URL (for container level access only)

This looks reasonably straight forward.  If you @maetthu and @memotype are willing to test it, I'll add the code?

**Comment by maetthu** [comment]
@ncw of course :) If needed, I can also provide you with access to a storage account.

**Comment by ncw** [comment]
I've made a first attempt at this:

https://beta.rclone.org/v1.40-019-gc76f1662-azureblob-auth/ (uploaded in 15-30 mins)

Run through `rclone config` and hopefully you'll work out what to do!

I haven't updated the docs yet.

I'd appreciate it if you could try the 3 auth methods - thanks!

**Comment by maetthu** [comment]
Authentication from connection string works when using account/key method, although the connection string needs to have the following format (copied from Azure Portal):

`DefaultEndpointsProtocol=https;AccountName=<account-name>;AccountKey=<account-key>;EndpointSuffix=core.windows.net`

rather than the one I wrote in my earlier comment, which was copied from Azure Storage Explorer (apparently, not all connection strings are created equally :roll_eyes:)

The other two connection methods (SAS in connection string and SAS URL) unfortunately don't work, I'll check if I can dig a bit deeper into the Go SDK to find out what's wrong, it might be just an encoding thing (although I followed the example here: https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string#create-a-connection-string-using-a-shared-access-signature with no luck)

* SAS in connection string reports following error:
```
2018/03/26 20:05:16 ERROR : : error reading source directory: storage: service returned error: StatusCode=400, ErrorCode=AccountRequiresHttps, ErrorMessage=The account being accessed does not support http.
RequestId:d89d2b4a-001e-0038-122c-c5161f000000
Time:2018-03-26T18:05:16.0525267Z, RequestInitiated=Mon, 26 Mar 2018 18:05:15 GMT, RequestId=d89d2b4a-001e-0038-122c-c5161f000000, API Version=, QueryParameterName=, QueryParameterValue=
2018/03/26 20:05:16 ERROR : Attempt 1/3 failed with 1 errors
2018/03/26 20:05:16 ERROR : : error reading source directory: storage: service returned error: StatusCode=400, ErrorCode=AccountRequiresHttps, ErrorMessage=The account being accessed does not support http.
RequestId:d89d2b5e-001e-0038-252c-c5161f000000
```

Although the error isn't really related to http vs https, if I enable http access to the storage account, the error changes to:
```
2018/03/26 20:24:33 ERROR : : error reading source directory: directory not found
2018/03/26 20:24:33 ERROR : Attempt 1/3 failed with 1 errors
```

* SAS url reports
```
2018/03/26 20:10:00 ERROR : : error reading source directory: Get ?comp=list&delimiter=%2F&include=metadata&maxresults=5000&restype=container: unsupported protocol scheme ""
2018/03/26 20:10:00 ERROR : Attempt 1/3 failed with 1 errors
```

Judging from this issue here https://github.com/Azure/azure-sdk-for-go/issues/464 it looks like there is  a separate go SDK for blob storage (https://github.com/Azure/azure-storage-blob-go) now apparently, where these issues might be addressed. I also found this pull request https://github.com/Azure/azure-sdk-for-go/pull/1055 which indicates that the storage part in the main go SDK will be moved away to dedicated packages, which is also noted in here: https://github.com/Azure/azure-sdk-for-go/blob/master/storage/README.md

I'm a bit confused atm what these two Go SDKs actually support or not, I'll try to find some more info about that in the next couple of days.

**Comment by memotype** [comment]
I appreciate the enthusiasm for this feature. @ncw Sorry I wasn't able to dig up the docs, and I'm definitely not a Go developer, just a simple sysadmin, so I won't be able to send a PR. We've currently implemented a work-around that's working for us, but using the rclone method with the SAS would definitely be preferred. Anyway, just wanted to let you know all know I wasn't just pulling a drive-by feature request. :P

**Comment by ncw** [comment]
@maetthu thanks for testing that :-)

> Authentication from connection string works

Hooray!

> SAS in connection string reports following error:

I note this in the docs - did you set https?

> DefaultEndpointsProtocol=[http|https];

This error

> error reading source directory: directory not found

Looks like it is almost working... 

If you wanted to do a bit more debugging with the code I posted, then I recommend using `-vv --dump bodies` which will show all the HTTP transactions - that usually helps a lot.

> SAS url reports

> error reading source directory: Get ?comp=list&delimiter=%2F&include=metadata&maxresults=5000&restype=container: unsupported protocol scheme ""

This looks like the URL is missing a protocol scheme, likely an `https:\\`  In fact it looks like it is missing the rest of the URL.  Did you put in something like "https://host.com/stuff` here?

`--dump bodies` may throw some light.

I've had interactions with the Azure Go devs and they indicated to me that the preferred package is now https://github.com/Azure/azure-storage-blob-go - however that package doesn't look particularly mature to me yet!  Also it is a lot of work swapping a backend implementation :-(

@memotype Thanks for the comment!  As far as this feature goes, happy to pull existing features out of the SDK, but I really don't have enough time to dive into the rabbit hole of the Microsoft auth system so really appreciate your help @maetthu.  Assuming we can get enough of this working to be useful, I'll ask for some help writing the docs too!

**Comment by ncw** [comment]
@memotype @maetthu just looking through branches I haven't merged and I found this one.

Does anyone want to help debug the bits which don't work, or shall I merge the connection string part (which does work) for v1.41?

**Comment by maetthu** [comment]
@ncw I'm sorry, I totally forgot about this :frowning_face: I'll do some debugging in the next days and see if I can shed some light on it. 

I'm not sure if merging the connection string part is a good idea though, since only the connection string with the account key works atm and this is already covered through the current method. Supporting connection strings but not all variations might just lead to more confusions w/o a real benefit. I guess it's better to wait until at least SAS through connection string also works.

**Comment by maetthu** [comment]
Here we go - I did some testing and digging and managed to get some more things to work:

* Fails: Config using account/key connection string from Azure Storage Explorer:

```
[azure]
type = azureblob
connection_string = DefaultEndpointsProtocol=https;AccountName=MYSTORAGEACCOUNT;AccountKey=KEY;BlobEndpoint=https://MYSTORAGEACCOUNT.blob.core.windows.net/;QueueEndpoint=https://MYSTORAGEACCOUNT.queue.core.windows.net/;TableEndpoint=https://MYSTORAGEACCOUNT.table.core.windows.net/;FileEndpoint=https://MYSTORAGEACCOUNT.file.core.windows.net/;                                                                                                       
```

```
2018/04/24 17:31:39 DEBUG : rclone: Version "v1.40-019-gc76f1662-azureblob-authβ" starting with parameters ["./rclone" "-vvv" "--config" "config-connection-string-key.txt" "copy" "tmp/" "azure:container"]
2018/04/24 17:31:39 DEBUG : Using config file from "/home/maetthu/tmp/rclone/config-connection-string-key.txt"
2018/04/24 17:31:39 Failed to create file system for "azure:container": failed to make azure storage client from connection string: azure: base storage service url required
```

* Works: Config using account/key connection string from Azure Portal:

```
[azure]
type = azureblob
connection_string = DefaultEndpointsProtocol=https;AccountName=MYSTORAGEACCOUNT;AccountKey=KEY;EndpointSuffix=core.windows.net;
```

* Works: Config using account-level SAS, generated in Azure Portal. Again, Storage Explorer connection string does not work since it doesn't include the protocol and the SDK falls back to HTTP, which fails unless the account is configured to allow unencrypted connections:

```
[azure]
type = azureblob
connection_string = SharedAccessSignature=sv=2017-04-17&ss=b&srt=sco&sp=rwdl&st=2018-04-24T15%3A32%3A14Z&se=2018-04-25T15%3A32%3A14Z&sig=SIGNATURE;BlobEndpoint=https://MYSTORAGEACCOUNT.blob.core.windows.net/;                 
```

* Fails: Config using an SAS-URL

```
[azure]
type = azureblob
sas_url = https://MYSTORAGEACCOUNT.blob.core.windows.net/?sv=2017-07-29&ss=b&srt=sco&sp=rwdlac&se=2018-04-25T00:25:09Z&st=2018-04-24T16:25:09Z&spr=https&sig=SIGNATURE                                                                 
```

```
$ ./rclone --dump headers --dump bodies --dump auth --dump requests --dump responses -vvv --config config-connection-string-sas-url.txt copy tmp/ azure:container
2018/04/24 18:29:44 DEBUG : rclone: Version "v1.40-019-gc76f1662-azureblob-authβ" starting with parameters ["./rclone" "--dump" "headers" "--dump" "bodies" "--dump" "auth" "--dump" "requests" "--dump" "responses" "-vvv" "--config" "config-connection-string-sas-url.txt" "copy" "tmp/" "azure:container"]
2018/04/24 18:29:44 DEBUG : Using config file from "/home/maetthu/tmp/rclone/config-connection-string-sas-url.txt"
2018/04/24 18:29:44 INFO  : Azure container container: Modify window is 1ns
2018/04/24 18:29:44 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2018/04/24 18:29:44 DEBUG : HTTP REQUEST (req 0xc420322b00)
2018/04/24 18:29:44 DEBUG : 
2018/04/24 18:29:44 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2018/04/24 18:29:44 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2018/04/24 18:29:44 DEBUG : HTTP RESPONSE (req 0xc420322b00)
2018/04/24 18:29:44 DEBUG : Error: unsupported protocol scheme ""
2018/04/24 18:29:44 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2018/04/24 18:29:44 ERROR : : error reading destination directory: Get ?comp=list&delimiter=%2F&include=metadata&maxresults=5000&restype=container: unsupported protocol scheme ""
2018/04/24 18:29:44 INFO  : Azure container container: Waiting for checks to finish
2018/04/24 18:29:44 INFO  : Azure container container: Waiting for transfers to finish
2018/04/24 18:29:44 ERROR : Attempt 1/3 failed with 1 errors
```

TL/DR:
* Connection string using account key works (if you use the connection string format from Azure Portal)
* Connection string using SAS works as well (if you manually craft a connection string from an SAS generated in Azure Portal)
* Both account-level and container-level SAS-URL don't work - seems it can't parse the URL properly? A trace shows that no connection to Azure is made.

**Comment by zifeo** [comment]
Is there any blocker on this? This would definitely be great to have!

**Comment by maetthu** [comment]
@zifeo help testing it would be nice I guess. I couldn't get SAS-URLs to work, but since I'm neither very experienced with Go nor with Azure, I might have missed something. For a link to download the version where the feature is added, see https://github.com/ncw/rclone/issues/2118#issuecomment-375891719

**Comment by ncw** [comment]
Yes more testing please!  I need help with understanding how the Azure stuff works - I understand the API quite well but the myriad ways of authentication I find quite baffling ;-)

**Comment by ncw** [comment]
Here is a new version of the code.  The code is unchanged but this has updated Azure libraries which may make a difference!

https://beta.rclone.org/v1.42-006-gacf9078d-azureblob-auth/ (uploaded in 15-30 mins)

If someone could test the SAS URL feature I'd be grateful!  (Configure a new azureblob remote and you'll see the option).

**Comment by zifeo** [comment]
I did some testing for the SAS-URL but only got this for copy: `2018/06/20 22:18:13 ERROR : Attempt 1/3 failed with 0 errors and: Head http://.blob./CONTAINER/FILES: dial tcp: lookup .blob.: no such host
`. Not sure I did everything correct though... I have only the `sas_url` set to the full url given by Azure?

**Comment by ncw** [comment]
@zifeo Thanks for testing this.  I found an issue on the upstream webpage: https://github.com/Azure/azure-sdk-for-go/issues/959 which isn't wildly helpful - it just says

> Are you sure your SAS query params are correct? I tried this and it works for me using a SAS token generated via the Azure portal.

**Comment by zifeo** [comment]
The token was generated through the Azure Portal as well. Maybe any encoding issue?

**Comment by ncw** [comment]
I managed to figure out that this method only works for container based SAS URLs so you have to open the storage explorer and right click on the container - there you will see an option to make a SAS URL.

This is almost working - well listings are working.  I can't get it to download files.

https://beta.rclone.org/v1.42-012-gef7b0541-fix-2118-azureblob-auth/ (uploaded in 15-30 mins)

I have made an upstream issue as I think there is a problem there: https://github.com/Azure/azure-sdk-for-go/issues/2119

**Comment by maetthu** [comment]
It seems a fix was merged here: https://github.com/Azure/azure-sdk-for-go/pull/2123

I expanded your example with a quick'n dirty blob upload and delete, built against the latest branch of Azure SDK it works for me with an SAS URL created within Storage Explorer now: https://github.com/maetthu/azure-storage-playground

**Comment by ncw** [comment]
@maetthu great - thanks for testing :-)

I've pulled the `latest ` branch from the azure repo into the rclone branch and rebuilt - can you give this a go?  It seems to be working for me!

I'll wait until the Azure code hits master before merging it to rclone master most likely.

https://beta.rclone.org/branch/v1.42-016-g865b3303-fix-2118-azureblob-auth/ (uploaded in 15-30 mins)

**Comment by zifeo** [comment]
I can confirm it works well for container specific SAS urls, nice job! Do you have an upstream issue for service wide SAS urls?

**Comment by maetthu** [comment]
@zifeo I didn't test account level SAS _URLs_, but an account level SAS is already possible using a connection string, see the results of my tests for a previous version here https://github.com/ncw/rclone/issues/2118#issuecomment-384008230

**Comment by zifeo** [comment]
Okay, in this case, we should make explicit that account level SAS urls does not work and rather connection string should be used.

**Comment by ncw** [comment]
@zifeo it does say this in the help

			Help: "SAS URL for container level access only\n(leave blank if using account/key or connection string)",

Maybe it could be more obvious.

Here is a first stab at some documentation - updates and corrections appreciated :-)

## Authenticating with Azure Blob Storage

Rclone has 3 ways of authenticating with Azure Blob Storage:

### Account and Key

This is the most straight forward and least flexible way.  Just fill in the `account` and `key` lines and leave the rest blank.

### Connection string

This supports all the possible connection string variants.  Leave `account`, `key` and `sas_url` blank and put the connection string into the `connection_string` configuration parameter.

Use this method if using an account level SAS; the Azure Portal shows connection strings you can cut and paste.

### SAS URL

This only for a container level SAS URL - it does not work with an account level SAS URL. For account level SAS use the connection string method.

To get a container level SAS URL right click on a container in the Azure Blob explorer in the Azure portal.

You will only be able to use the container specified in the SAS URL with rclone, eg

    rclone ls azureblob:container

However these will not work

    rclone lsd azureblob:
    rclone ls azureblob:othercontainer

This would be useful for temporarily allowing third parties access to a single container or putting credentials into an untrusted environment.

**Comment by zifeo** [comment]
I might have gone too fast on it, sorry. The document seems all good to me!

**Comment by maetthu** [comment]
@ncw works here as well, using `sas_url` in config:

```
[azure]
type = azureblob
sas_url = https://ACCOUNT.blob.core.windows.net/CONTAINTER?sv=2017-07-29...
```

very nice :tada:

**Comment by ncw** [comment]
Excellent - thanks for testing :-) I'm going to leave this open for a while until the upstream fix hits the master branch, then I'll merge to master.

**Comment by ncw** [comment]
Apparently the upstream fix is due very soon so I'll hold off for a moment.

> We're planning to release v18 first week of July (hopefully the 2nd), and that release will have the fix.

**Comment by ncw** [comment]
I have merged this to master now - thank you all for your help testing.

https://beta.rclone.org/v1.42-022-g03c5b823/ (uploaded in 15-30 mins)

**Comment by zifeo** [comment]
Thanks, when is the release planned?

**Comment by maetthu** [comment]
@ncw very nice, thank you :slightly_smiling_face:

**Comment by ncw** [comment]
The release is planned for 28th July.  No guarantees though!

**Comment by ncw** [comment]
Note that thanks to the hard work by @sandeepkru we've updated to use the newer SDK.

This is the recommended SDK for use with azure blob.

It has the advantage that SAS URLs now work properly both for account and container.

It has the disadvantage that we've had to drop connection string support.  Apparently connection strings are no longer going to be supported by Microsoft which is why the new SDK doesn't support them.  I don't view this as a big loss since we only just put that feature in, but it may affect people on this issue, so I apologise if this causes you any inconvenience.

You'll find this in

https://beta.rclone.org/v1.42-035-g99201f8b/ (uploaded in 15-30 mins)

and [the latest beta](https://beta.rclone.org) from now on.
